File: bench/window/window.go
// Sliding-window, Go: a ring of strings; overwriting a slot makes the evicted
// string garbage (GC reclaims it), so peak RSS tracks the window + GC headroom.
package main
import ("fmt"; "strconv")
func main() {
    const n, w = 2000000, 50000
    ring := make([]string, w)
    acc := 0
    for i := 0; i < n; i++ {
        ring[i%w] = "rec" + strconv.Itoa(i%100000)
        acc += len(ring[i%w])
    }
    fmt.Println(acc)
}

File: bench/window/window.c
/* Sliding-window, C: a ring of strdup'd strings; the evicted slot is freed, so
 * peak RSS tracks the WINDOW (W), not the stream (N). Same checksum as the hier port. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
int main(void) {
    long n = 2000000, w = 50000, acc = 0;
    char **ring = calloc((size_t)w, sizeof(char *));
    for (long i = 0; i < n; i++) {
        char buf[32];
        snprintf(buf, sizeof buf, "rec%ld", i % 100000);
        long s = i % w;
        free(ring[s]);
        ring[s] = strdup(buf);
        acc += (long)strlen(ring[s]);
    }
    printf("%ld\n", acc);
    return 0;
}
